orttraining/orttraining/core/optimizer/memory_alleviation.cc

```diff
@@ -0,0 +1,460 @@
+      }
+    }
+  }
+
+  return Status::OK();
+}
+
+Status MemoryAlleviation::RegisterRecomputableIntermediateOps() {
+  recomputable_intermediate_op_crawler_map_["Where"] =
+      [](const Graph& graph, const Node& node,
+         InlinedVector<memory_alleviation::NodeOutputPort>& input_node_output_args) -> bool {
+    const Node* data_true_node = graph.GetProducerNode(node.InputDefs()[1]->Name());
+    if (!data_true_node) {
+      // input is graph inptus.
+      return true;
+    }
+
+    size_t producer_output_index = 0;
+    for (size_t i = 0; i < data_true_node->OutputDefs().size(); ++i) {
+      if (data_true_node->OutputDefs()[i]->Name().compare(node.InputDefs()[1]->Name()) == 0) {
+        producer_output_index = i;
+        break;
+      }
+    }
+
+    // False condition be a scalar constant.
+    if (!graph.GetConstantInitializer(node.InputDefs()[2]->Name(), true)) {
+      return false;
+    }
+
+    input_node_output_args.push_back(
+        std::make_pair(std::move(data_true_node), static_cast<int>(producer_output_index)));
+    return true;
+  };
+
+  return Status::OK();
+}
+
+Status MemoryAlleviation::PrepareForTransformation(const Graph& graph,
+                                                   ActivationUsedMap& fw_op_output_arg_used_map,
+                                                   InlinedHashMap<NodeIndex, bool>& is_forward_op_map) const {
+  fw_op_output_arg_used_map.clear();
+  is_forward_op_map.clear();
+  InlinedHashMap<std::string, InlinedHashSet<const Node*>> node_arg_name_to_consumer_node_map;
+
+  GraphViewer graph_viewer(graph);
+  const auto& node_ids = graph_viewer.GetNodesInTopologicalOrder();
+
+  bool is_forward = true;
+  size_t dropout_node_index = 0;
+  for (size_t i = 0; i < node_ids.size(); ++i) {
+    const Node& node = *graph.GetNode(node_ids[i]);
+    if (dropout_alleviation_type_ == memory_alleviation::AlleviationType::Recompute &&
+        ((node.OpType() == "Dropout" || node.OpType() == "BitmaskDropout"))) {
+      // We assign seed for all dropout nodes in case duplicated/recomputed node use different seeds with original ones.
+      const_cast<Node&>(node).AddAttribute("seed", static_cast<int64_t>(dropout_node_index + utils::GetRandomSeed()));
+      dropout_node_index++;
+    }
+
+    if (is_forward) {
+      for (auto& output_arg : node.OutputDefs()) {
+        for (auto& consumer_node : graph.GetConsumerNodes(output_arg->Name())) {
+          node_arg_name_to_consumer_node_map[output_arg->Name()].insert(consumer_node);
+        }
+      }
+    }
+
+    is_forward_op_map[node.Index()] = is_forward;
+    if (node.OpType() == "YieldOp") {
+      is_forward = false;
+    }
+  }
+
+  for (auto& kv : node_arg_name_to_consumer_node_map) {
+    for (auto& consumer_node : kv.second) {
+      if (is_forward_op_map[consumer_node->Index()]) {
+        fw_op_output_arg_used_map[kv.first].first = true;
+      } else {
+        fw_op_output_arg_used_map[kv.first].second = true;
+      }
+    }
+  }
```

Comment:
I did not find a good way to merge them, because the first pass need go through all nodes and know it is fw op or not, then the second loop will loop up those stats for each node's consumers, while I refined the loops as you suggested.